v2/src/clients/SugarShareClient.ts

```diff
@@ -0,0 +1,101 @@
+import axios, { AxiosInstance } from 'axios';
+
+import settings from '../settings';
+import Auth, { AuthenticationError } from 'libs/auth';
+import { Callback } from 'libs/types';
+
+interface PresignedUrlBody {
+  title: string;
+  fileType: string;
+  sizeBytes: number;
+}
+
+interface PresignedUrlResponse {
+  uuid: string;
+  presignedUrl: string;
+}
+
+class SugarShareClient {
+  private readonly client: AxiosInstance;
+  private readonly auth: Auth;
+  private readonly abortController: AbortController;
+
+  constructor() {
+    this.abortController = new AbortController();
+    this.auth = new Auth();
+    this.client = axios.create({
+      baseURL: `https://${settings.apiDomainName}`,
+      signal: this.abortController.signal,
+    });
+  }
+
+  /**
+   * Initiate upload by fetching a presigned S3 PUT URL
+   */
+  private async getPresignedUrl(file: File): Promise<PresignedUrlResponse> {
+    const body: PresignedUrlBody = {
+      title: file.name,
+      fileType: file.type,
+      sizeBytes: file.size,
+    };
+
+    let authorizationToken = '';
+    try {
+      await this.auth.load();
+      authorizationToken = await this.auth.getAuthorizationToken();
+    } catch (error) {
+      if (error instanceof AuthenticationError && error.message.match(/Cannot find tokens in storage/i)) {
+        // Skip and a 401 will be returned from the API and displayed
+      } else {
+        throw error;
+      }
+    }
+
+    return (await this.client.post(
+      '/init',
+      body,
+      {
+        headers: {
+          ...(authorizationToken.length && { Authorization: `Bearer ${authorizationToken}` }),
+        },
+      },
+    )).data;
+  }
+
+  /**
+   * PUT file into bucket using a presigned URL
+   */
+  private async putFile(file: File, presignedUrl: PresignedUrlResponse['presignedUrl'], handleProgress: Callback<number>) {
+    await this.client.put(presignedUrl, file, {
+      onUploadProgress: (event) => {
+        const progress = Math.round((event.loaded * 100) / event.total);
+        handleProgress(progress);
+      },
+      headers: {
+        'Content-Type': file.type,
+      },
+    });
+  }
+
+  /**
+   * Upload file and return a shareable link
+   */
+  public async upload(file: File, handleProgress: Callback<number>): Promise<string> {
+    const { presignedUrl } = await this.getPresignedUrl(file);
+    await this.putFile(file, presignedUrl, handleProgress);
+
+    const url = new URL(presignedUrl);
+    return `${url.hostname}${url.pathname}`;
+  }
+
+  /**
+   * Cancel all HTTP(s) requests performed by the client
+   */
+  public cancel(): void {
+    this.abortController.abort();
+  }
+}
+
+const SugarShare = new SugarShareClient();
```

Comment:
I believe this wont be a problem, I have done that for most apis but Im not sure  if the SugarShare API support this

Reply:
Why not `export default SugarShareClient`?

Or shorter with `export default class SugarShareClient`.

Reply:
I want to void OOP inside React as possible, imagine if the API Client to "remove a file" was something like this

const SugarShare = new SugarShareClient(fileId)

SugarShare.deleteFile() // the fileId here is a private property and is being handled inside

The thing is that now I have a "problem" that is so unnecessary, first Ill have to tight my Client to be Similar to the API that it doesn't make a lot of sense, for example

Id need to have a component that take his id as props and do something like this:

```js
const MyComponentWithIdAsProps = ({myId}) => {
const SugarShare = new SugarShareApi(myId) 
// this also means that EVERYTIME my component re-render this API will be initialized and instantiated again
// my component can have a lot of state related to UI only, what if it gets "red" for some reason?  it will also instatiate the API once again
// what if it toggles when something happens (idk just an example)

IMO this all would make a bit more complicated when creating or handling components to also take in consideration the API, where I think a best approach would be something like
}
```


```js
const MyDumbComponent = (pros) => {
// it will take all the needed props, including even maybe the id, but the only thing this component does is to care about the UI aspect of the thing, if I want this to reflect something related to the API, I can always pass it as props like "loading, error". etc
}
And the parent would handle the "side-effect" of adding a new item on the upload list for example:

Every time we add a new item:
1. We add it to the list state 
2. WE make an API request with the data

If the API fails we can then use .catch and dispatch a new state change that will show to the user that the upload file -> and this is the important part, its not the component that handle its own request, or even its parent, the parent only handle the side effect of the API call

```

Reply:
This explains way better than me: https://reactjs.org/docs/thinking-in-react.html